cpp/test/sparse/preprocess_csr.cu

```diff
@@ -0,0 +1,209 @@
+                                                           csr_in.get_elements().size());
+  auto rows   = raft::make_device_vector<T1, int64_t>(handle, values.size());
+
+  raft::sparse::convert::csr_to_coo<T1>(
+    indptr.data_handle(), num_rows, rows.data_handle(), rows.size(), stream);
+
+  raft::copy(h_rows.data_handle(), rows.data_handle(), rows.size(), stream);
+  raft::copy(h_cols.data_handle(), indices.data_handle(), cols_size, stream);
+  raft::copy(h_elems.data_handle(), values.data_handle(), values.size(), stream);
+  raft::util::preproc_coo<T1, T2>(
+    handle, h_rows.view(), h_cols.view(), h_elems.view(), results, num_rows, num_cols, tf_idf);
+}
+
+template <typename Type_f, typename Index_>
+struct SparsePreprocessInputs {
+  int n_rows;
+  int n_cols;
+  int nnz_edges;
+};
+
+template <typename Type_f, typename Index_>
+class SparsePreprocessCSR
+  : public ::testing::TestWithParam<SparsePreprocessInputs<Type_f, Index_>> {
+ public:
+  SparsePreprocessCSR()
+    : params(::testing::TestWithParam<SparsePreprocessInputs<Type_f, Index_>>::GetParam()),
+      stream(resource::get_cuda_stream(handle))
+  {
+  }
+
+ protected:
+  void SetUp() override {}
+
+  void Run(bool bm25_on)
+  {
+    cudaStream_t stream = raft::resource::get_cuda_stream(handle);
+
+    int num_rows = pow(2, params.n_rows);
+    int num_cols = pow(2, params.n_cols);
+
+    auto rows    = raft::make_device_vector<Index_, int64_t>(handle, params.nnz_edges);
+    auto columns = raft::make_device_vector<Index_, int64_t>(handle, params.nnz_edges);
+    auto values  = raft::make_device_vector<Type_f, int64_t>(handle, params.nnz_edges);
+    auto mask    = raft::make_device_vector<Index_, int64_t>(handle, params.nnz_edges);
+
+    raft::util::create_dataset<Index_, Type_f>(
+      handle, rows.view(), columns.view(), values.view(), 5, params.n_rows, params.n_cols);
+    int non_dupe_nnz_count = raft::util::get_dupe_mask_count<Index_, Type_f>(
+      handle, rows.view(), columns.view(), values.view(), mask.view());
+
+    auto rows_nnz    = raft::make_device_vector<Index_, int64_t>(handle, non_dupe_nnz_count);
+    auto columns_nnz = raft::make_device_vector<Index_, int64_t>(handle, non_dupe_nnz_count);
+    auto values_nnz  = raft::make_device_vector<Type_f, int64_t>(handle, non_dupe_nnz_count);
+    raft::util::remove_dupes<Index_, Type_f>(handle,
+                                             rows.view(),
+                                             columns.view(),
+                                             values.view(),
+                                             mask.view(),
+                                             rows_nnz.view(),
+                                             columns_nnz.view(),
+                                             values_nnz.view(),
+                                             num_rows);
+    auto rows_csr = raft::make_device_vector<Index_, int64_t>(handle, non_dupe_nnz_count);
+    raft::sparse::convert::sorted_coo_to_csr(
```

Comment:
I noticed this might be helpful in resolving this error(maybe we can check if the allocated memory size is correct or the `indices` staff has no invalid ones):
```shell
========= Invalid __global__ write of size 4 bytes
=========     at void cub::CUB_200500_890_NS::DeviceScanKernel<cub::CUB_200500_890_NS::DeviceScanPolicy<int, thrust::plus<void>>::Policy900, thrust::device_ptr<int>, thrust::device_ptr<int>, cub::CUB_200500_890_NS::ScanTileState<int, (bool)1>, thrust::plus<void>, cub::CUB_200500_890_NS::detail::InputValue<int, int *>, int, int>(T2, T3, T4, int, T5, T6, T7)+0x1b30
=========     by thread (8,0,0) in block (0,0,0)
=========     Address 0x7c6486200e20 is out of bounds
```